pkg/resource/clusteroperator_test.go

```diff
@@ -0,0 +1,192 @@
+	labels "k8s.io/apimachinery/pkg/labels"
+)
+
+type deployLister struct {
+	deploys   map[string]appsv1.Deployment
+	failOnGet bool
+}
+
+func (d deployLister) Get(name string) (*appsv1.Deployment, error) {
+	if d.failOnGet {
+		return nil, &kerror.StatusError{
+			ErrStatus: metav1.Status{
+				Reason: metav1.StatusReasonInternalError,
+			},
+		}
+	}
+
+	deploy, ok := d.deploys[name]
+	if !ok {
+		return nil, &kerror.StatusError{
+			ErrStatus: metav1.Status{
+				Reason: metav1.StatusReasonNotFound,
+			},
+		}
+	}
+	return &deploy, nil
+}
+
+func (d deployLister) List(selector labels.Selector) ([]*appsv1.Deployment, error) {
+	return nil, nil
+}
+
+func TestSyncVersions(t *testing.T) {
+	lister := deployLister{}
+	for _, tt := range []struct {
+		name         string
+		environ      map[string]string
+		deploys      map[string]appsv1.Deployment
+		versions     []cfgapi.OperandVersion
+		modified     bool
+		expectsError bool
+		failOnGet    bool
+	}{
+		{
+			name:         "no env and no deployment",
+			modified:     false,
+			expectsError: false,
+		},
+		{
+			name:         "api error",
+			modified:     false,
+			expectsError: true,
+			failOnGet:    true,
+		},
+		{
+			name:         "only env",
+			modified:     true,
+			expectsError: false,
+			environ: map[string]string{
+				"RELEASE_VERSION": "1",
+			},
+			versions: []cfgapi.OperandVersion{
+				{
+					Name:    "operator",
+					Version: "1",
+				},
+			},
+		},
+		{
+			name:         "a not ready deployment",
+			modified:     false,
+			expectsError: false,
+			deploys: map[string]appsv1.Deployment{
+				imregv1.ImageRegistryName: {
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							imregv1.VersionAnnotation: "2",
+						},
+					},
+				},
+			},
+		},
+		{
+			name:         "a ready deployment",
+			modified:     true,
+			expectsError: false,
+			deploys: map[string]appsv1.Deployment{
+				imregv1.ImageRegistryName: {
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							imregv1.VersionAnnotation: "2",
+						},
+					},
+					Status: appsv1.DeploymentStatus{
+						AvailableReplicas: 1,
+					},
+				},
+			},
+			versions: []cfgapi.OperandVersion{
+				{
+					Name:    "image-registry",
+					Version: "2",
+				},
+			},
+		},
+		{
+			name:         "env and a ready deployment",
+			modified:     true,
+			expectsError: false,
+			environ: map[string]string{
+				"RELEASE_VERSION": "1",
+			},
+			deploys: map[string]appsv1.Deployment{
+				imregv1.ImageRegistryName: {
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							imregv1.VersionAnnotation: "2",
+						},
+					},
+					Status: appsv1.DeploymentStatus{
+						AvailableReplicas: 1,
+					},
+				},
+			},
+			versions: []cfgapi.OperandVersion{
+				{
+					Name:    "operator",
+					Version: "1",
+				},
+				{
+					Name:    "image-registry",
+					Version: "2",
+				},
+			},
+		},
```

Comment:
Add a realistic upgrade scenario

1. Operator is at version 2
2. image-registry is still at version 1

Reply:
Done.

Reply:
Done.